crates/relayer-framework-test/src/relayer_mock/contexts/chain.rs

```diff
@@ -0,0 +1,302 @@
+
+        // Update the current_state of the Chain
+        self.update_current_state(new_state)?;
+
+        Ok(())
+    }
+
+    pub fn build_send_packet(
+        &self,
+        client_id: ClientId,
+        channel_id: ChannelId,
+        port_id: PortId,
+        timeout_height: MockHeight,
+        timeout_timestamp: MockTimestamp,
+        sequence: Sequence,
+    ) -> PacketKey {
+        PacketKey::new(
+            client_id,
+            channel_id,
+            port_id,
+            sequence,
+            timeout_height,
+            timeout_timestamp,
+        )
+    }
+
+    /// Update the chain's current_state and past_chain_states at the same time to insure
+    /// they are always synchronized.
+    /// The MockChain must have a one and only one ChainState at every height.
+    fn update_current_state(&self, state: State) -> Result<(), Error> {
+        let latest_height = self.get_latest_height()?;
+        let new_height = latest_height.increment();
+        let mut locked_current_state = self.current_state.acquire_mutex()?;
+        locked_current_state.insert(new_height.clone(), state);
+
+        // After inserting the new state in the current_state, update the past_chain_states
+        // at the given height.
+        let mut locked_past_chain_states = self.past_chain_states.acquire_mutex()?;
+        locked_past_chain_states.insert(new_height, locked_current_state.clone());
+
+        Ok(())
+    }
+
+    /// If the message is a `SendPacket`, update the received packets,
+    /// and add a `RecvPacket` event to the returned array of events.
+    /// If the message is an `AckPacket`, update the received acknowledgment
+    /// packets.
+    /// If the message is an `UpdateClient` update the consensus state.
+    /// When a RecvPacket and AckPacket are received, verify that the client
+    /// state has respectively sent the message and received the message.
+    pub fn process_messages(&self, messages: Vec<MockMessage>) -> Result<Vec<Vec<Event>>, Error> {
+        let mut res = vec![];
+        for m in messages {
+            match m {
+                MockMessage::RecvPacket(receiver, h, p) => {
+                    let client_consensus =
+                        self.query_consensus_state_at_height(receiver.clone(), h.clone())?;
+                    let state = client_consensus.get(&h).ok_or_else(|| {
+                        Error::no_consensus_state_at_height(self.name().to_string(), h.0)
+                    })?;
+                    if !state.check_sent(&p.port_id, &p.channel_id, &p.sequence) {
+                        return Err(Error::generic(eyre!("chain `{}` got a RecvPacket, but client `{}` state doesn't have the packet as sent", self.name(), receiver)));
+                    }
+                    // Check that the packet is not timed out. Current height < packet timeout height.
+                    self.receive_packet(p)?;
```

Comment:
Should the validation logic above be implemented inside this function, so that the `process_messages` logic can be kept simple?